### PR TITLE
Improve upgrade test to run multiple pods

### PR DESCRIPTION
**Pull request by meldafrawi**
Longhorn/Longhorn#1391

### CODE REVIEW

File: manager/integration/tests/common.py [hunk 1]
```diff
@@ -3287,15 +3287,20 @@ def expand_attached_volume(client, volume_name):
 
 def prepare_pod_with_data_in_mb(
         client, core_api, pod_make, volume_name, volume_size=str(1*Gi),
-        data_path="/data/test", data_size_in_mb=DATA_SIZE_IN_MB_1):  # NOQA:
+        data_path="/data/test", data_size_in_mb=DATA_SIZE_IN_MB_1,
+        add_liveness_prope=True):  # NOQA:
```
Comment (yasker) [inline manager/integration/tests/common.py:3291]: The title of the commit is wrong. This part is not a fix for a bug.

File: manager/integration/tests/test_upgrade.py [hunk 2]
```diff
@@ -31,7 +48,7 @@ def longhorn_upgrade(image_tag):
 
 
 @pytest.mark.upgrade
-def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
+def test_upgrade(upgrade_image_tag, volume_name, pod_make, statefulset, storage_class):  # NOQAA
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:51]: "NOQAA"? How did this pass flake8...

File: manager/integration/tests/test_upgrade.py [hunk 3]
```diff
@@ -31,7 +48,7 @@ def longhorn_upgrade(image_tag):
...
     """
     Test Longhorn upgrade
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:54]: Add prerequisite. 

As I mentioned, `auto-salvage` should be disabled here.

File: manager/integration/tests/test_upgrade.py [hunk 4]
```diff
@@ -45,39 +62,111 @@ def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
...
+        name=statefulset_name,
+        namespace='default',
+        body={
+            'spec': {
+                'replicas': replicas
+            }
+        })
 
-    volume = client.by_id_volume(volume_name)
+    delete_and_wait_pod(core_api, pod_name)
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:125]: Have we checked the pod's data?

File: manager/integration/tests/test_upgrade.py [hunk 5]
```diff
@@ -45,39 +62,111 @@ def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
...
 
-    data = write_volume_random_data(volume)
+    for sspod_info in statefulset_pod_info:
+        sspod = core_api.read_namespaced_pod(name=sspod_info['pod_name'],
+                                             namespace='default')
+        assert \
+            sspod.status.container_statuses[0].restart_count == 0
 
-    assert longhorn_upgrade(upgrade_image_tag)
+    statefulset['spec']['replicas'] = replicas = 0
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:113]: Have we checked the data of statefulset?

File: manager/integration/tests/test_upgrade.py [hunk 6]
```diff
@@ -45,39 +62,111 @@ def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
...
+        body={
+            'spec': {
+                'replicas': replicas
+            }
+        })
 
-    volume = client.by_id_volume(volume_name)
+    delete_and_wait_pod(core_api, pod_name)
 
+    volume = client.by_id_volume(volume_name)
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:127]: Check the volume data?

File: manager/integration/tests/test_upgrade.py (hunk 3 above)
Comment (yasker) [reply manager/integration/tests/test_upgrade.py:54]: Also, update the skeleton.

File: manager/integration/tests/test_upgrade.py [hunk 7]
```diff
@@ -45,39 +62,111 @@ def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
...
+    apps_api.patch_namespaced_stateful_set(
+        name=statefulset_name,
+        namespace='default',
+        body={
+            'spec': {
+                'replicas': replicas
+            }
+        })
+
+    wait_statefulset(statefulset)
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:159]: We need to check the statefulset data. (do we have a way to do it?)

File: manager/integration/tests/test_upgrade.py (hunk 5 above)
Comment (yasker) [reply manager/integration/tests/test_upgrade.py:113]: We need to make sure data is correct after the manager upgrade but before the engine upgrade.

After the engine upgrade, write more data to the volume, then upgrade the engine.

**Comment by yasker** [comment]
The titles of two commits need to be updated as well. It doesn't indicate what's the changes and why we made them.

File: manager/integration/tests/common.py (hunk 1 above)
Comment (meldafrawi) [reply manager/integration/tests/common.py:3302]: fixed

File: manager/integration/tests/test_upgrade.py (hunk 2 above)
Comment (meldafrawi) [reply manager/integration/tests/test_upgrade.py]: fixed

File: manager/integration/tests/test_upgrade.py (hunk 3 above)
Comment (meldafrawi) [reply manager/integration/tests/test_upgrade.py:59]: Updated

File: manager/integration/tests/test_upgrade.py (hunk 4 above)
Comment (meldafrawi) [reply manager/integration/tests/test_upgrade.py:178]: Updated it to check data before/after engine upgrade

File: manager/integration/tests/test_upgrade.py (hunk 5 above)
Comment (meldafrawi) [reply manager/integration/tests/test_upgrade.py:166]: fixed

File: manager/integration/tests/test_upgrade.py (hunk 7 above)
Comment (meldafrawi) [reply manager/integration/tests/test_upgrade.py:212]: yes, fixed

File: manager/integration/tests/test_upgrade.py [hunk 8]
```diff
@@ -31,53 +53,177 @@ def longhorn_upgrade(image_tag):
 
 
 @pytest.mark.upgrade
-def test_upgrade(upgrade_image_tag, volume_name):  # NOQA
+def test_upgrade(upgrade_image_tag, volume_name, pod_make, statefulset, storage_class):  # NOQA
     """
     Test Longhorn upgrade
 
+    Prerequisite:
+      - Disable Auto Salvage Setting
```
Comment (yasker) [inline manager/integration/tests/test_upgrade.py:61]: Does this test case has a side-effect of disabling the auto salvage setting?

Comment (meldafrawi) [reply manager/integration/tests/test_upgrade.py:62]: added `reset_settings` to the end of the test.

Comment (yasker) [reply manager/integration/tests/test_upgrade.py:62]: That won't work if the test failed.

**Pull request merged**